src/vm/stablehashcode.cpp

```diff
@@ -0,0 +1,137 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "stablehashcode.h"
+#include "typehashingalgorithms.h"
+
+// DIFFERENT FROM CORERT: Hashing code currently different from CoreRT
+
+DWORD GetVersionResilientTypeHashCode(IMDInternalImport *pMDImport, mdExportedType token)
+{
+    _ASSERTE(TypeFromToken(token) == mdtTypeDef ||
+        TypeFromToken(token) == mdtTypeRef ||
+        TypeFromToken(token) == mdtExportedType);
+    _ASSERTE(!IsNilToken(token));
+
+    LPCUTF8 szNamespace;
+    LPCUTF8 szName;
+    DWORD hashcode = 0;
+
+    while (true)
+    {
+        if (IsNilToken(token))
+            return hashcode;
+
+        switch (TypeFromToken(token))
+        {
+        case mdtTypeDef:
+            if (FAILED(pMDImport->GetNameOfTypeDef(token, &szName, &szNamespace)))
+                ThrowHR(COR_E_BADIMAGEFORMAT);
+            if (FAILED(pMDImport->GetNestedClassProps(token, &token)))
+                token = mdTokenNil;
+            break;
+
+        case mdtTypeRef:
+            if (FAILED(pMDImport->GetNameOfTypeRef(token, &szNamespace, &szName)))
+                ThrowHR(COR_E_BADIMAGEFORMAT);
+            if (FAILED(pMDImport->GetResolutionScopeOfTypeRef(token, &token)))
+                token = mdTokenNil;
+            break;
+
+        case mdtExportedType:
+            if (FAILED(pMDImport->GetExportedTypeProps(token, &szNamespace, &szName, &token, NULL, NULL)))
+                ThrowHR(COR_E_BADIMAGEFORMAT);
+            break;
+
+        default:
+            return hashcode;
+        }
+
+        hashcode ^= ComputeNameHashCode(szNamespace, szName);
+    }
+}
+
+#ifndef DACCESS_COMPILE
+DWORD GetVersionResilientTypeHashCode(TypeHandle type)
+{
+    if (!type.IsTypeDesc())
+    {
+        MethodTable *pMT = type.AsMethodTable();
+
+        DWORD hashcode = 0;
+        if (pMT->IsArray())
+        {
+            return ComputeArrayTypeHashCode(GetVersionResilientTypeHashCode(pMT->GetApproxArrayElementTypeHandle()), pMT->GetRank());
+        }
+        else if (!IsNilToken(pMT->GetCl()))
```

Comment:
We should always get a nice type with metadata here. Do we really need the IsArray check (arrays should come as typedescs here) and the null token check?